src/Adapter/Image/Uploader/ProductImageUploader.php

```diff
@@ -0,0 +1,119 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Image\Uploader;
+
+use ErrorException;
+use Hook;
+use Image;
+use PrestaShop\PrestaShop\Adapter\Image\Exception\CannotUnlinkImageException;
+use PrestaShop\PrestaShop\Adapter\Product\ProductImagePathFactory;
+use PrestaShop\PrestaShop\Core\Configuration\UploadSizeConfigurationInterface;
+
+//@todo: do we really need an interface? depends if we use it in controller or in command handler
+class ProductImageUploader extends AbstractImageUploader
+{
+    /**
+     * @var UploadSizeConfigurationInterface
+     */
+    private $uploadSizeConfiguration;
+
+    /**
+     * @var ProductImagePathFactory
+     */
+    private $productImagePathFactory;
+
+    /**
+     * @var int
+     */
+    private $contextShopId;
+
+    /**
+     * @param UploadSizeConfigurationInterface $uploadSizeConfiguration
+     * @param ProductImagePathFactory $productImagePathFactory
+     * @param int $contextShopId
+     */
+    public function __construct(
+        UploadSizeConfigurationInterface $uploadSizeConfiguration,
+        ProductImagePathFactory $productImagePathFactory,
+        int $contextShopId
+    ) {
+        $this->uploadSizeConfiguration = $uploadSizeConfiguration;
+        $this->productImagePathFactory = $productImagePathFactory;
+        $this->contextShopId = $contextShopId;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function upload(Image $image, string $filePath): void
+    {
+        $this->productImagePathFactory->createDestinationDirectory($image);
+
+        //@todo: this will unlink the image. Can we trust that the $filePath is in temp?
+        $this->uploadFromTemp($filePath, $this->productImagePathFactory->getBasePath($image, true));
+        $this->generateDifferentSizeImages($this->productImagePathFactory->getBasePath($image, false), 'products');
+
+        Hook::exec('actionWatermark', ['id_image' => (int) $image->id, 'id_product' => (int) $image->id_product]);
+        //@todo: moved multishop association from here to Repository when Image objModel is created
+        $this->deleteCachedImages($image);
+    }
+
+    /**
+     * @param Image $image
+     *
+     * @throws CannotUnlinkImageException
+     */
+    private function deleteCachedImages(Image $image): void
+    {
+        $productId = (int) $image->id_product;
+
+        $cachedImages = [
+            $this->productImagePathFactory->getCachedCover($productId, $this->contextShopId),
+            $this->productImagePathFactory->getCachedThumbnail($productId),
+        ];
+
+        foreach ($cachedImages as $cachedImage) {
+            if (file_exists($cachedImage)) {
+                try {
+                    unlink($cachedImage);
+                } catch (ErrorException $e) {
+                    //@todo: do we really need to fail on cached images deletion? It was suppressed in legacy
```

Comment:
Yes. Failing and throwing an exception is valid for the domain code.
At Controller level, shop user can decide to ignore this but the domain code should not question "should I ignore this ?"

Domain code should always send error signals to the upper layer, upper layer decides what to acknowledge and what to ignore